### PR TITLE
Fix transmute on ByValPair

### LINKED ISSUE #227: transmute on `ByValPair` is wrong if field sizes of source and target don't match
```rust
    let x: Option<Box<[u8]>> = unsafe {
        let z = std::intrinsics::add_with_overflow(0usize, 0usize);
        std::mem::transmute::<(usize, bool), Option<Box<[u8]>>>(z)
    };
    let y = &x;
```

The allocation pointed to by `y` now contains all zeros, when it should be 9 zeros and 7 undef bytes.

### PR DESCRIPTION
**Pull request by RalfJung**
Fixes #227

### CODE REVIEW

**Comment by oli-obk** [comment]
A related test fails on i686

**Comment by RalfJung** [comment]
Oh, right, this test doesn't really work that way on 32-bit systems.

**Comment by RalfJung** [comment]
Tests succeed now.

File: src/terminator/intrinsic.rs [hunk 1]
```diff
@@ -395,16 +395,10 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
...
-                if dest_align < src_align {
-                    let ptr = self.force_allocation(dest)?.to_ptr()?;
-                    self.memory.mark_packed(ptr, size);
-                    self.write_value_to_ptr(arg_vals[0], PrimVal::Ptr(ptr), dest_ty)?;
-                } else {
-                    self.write_value(arg_vals[0], dest, dest_ty)?;
-                }
+                let ptr = self.force_allocation(dest)?.to_ptr()?;
+                self.memory.mark_packed(ptr, size);
+                self.write_value_to_ptr(arg_vals[0], PrimVal::Ptr(ptr), src_ty)?;
```
Comment (oli-obk) [inline src/terminator/intrinsic.rs]: Oh wow! That was a subtle bug

Comment (RalfJung) [inline src/terminator/intrinsic.rs]: Yeah, it was using the wrong type for writing.

I wonder if there is anything we can do to make such bugs (using the wrong type) surface easier... but I can't think of anything.

Comment (oli-obk) [inline src/terminator/intrinsic.rs]: We can fuzz miri by randomly producing safe code xD

**Pull request merged**